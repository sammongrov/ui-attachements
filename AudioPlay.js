import React, { Component } from 'react';
import { View, Text, Slider, TouchableOpacity, Alert } from 'react-native';
import PropTypes from 'prop-types';
import AppUtil from '@mongrov/utils'; 
import { Player } from 'react-native-audio-toolkit';
import { Colors } from '@ui/theme_default';
import { styles } from 'react-native-theme';
import { Icon, FontAwesomeIcon } from '@ui/components';

export default class AudioPlay extends Component {
  constructor() {
    super();

    this.state = {
      playButtonDisabled: false,
      progress: 0,
      playTime: 0,
      duration: 0,
      playing: false,
      loadingAudio: false,
      playerExists: false,
    };
  }

  componentWillMount() {
    this.player = null;
    this.lastSeek = 0;
    this._getProgressTimer();
  }

  componentWillUnmount() {
    if (this._progressInterval) {
      clearInterval(this._progressInterval);
      this._progressInterval = null;
    }
    if (this._playInterval) {
      clearInterval(this._playInterval);
      this._playInterval = null;
    }
    this._deletePlayer();
  }

  getTimerString(playTime) {
    const recT = playTime / 1000;
    const recM = Math.floor(recT / 60);
    const recS = Math.round(recT - recM * 60);
    return `${recM < 10 ? '0' : ''}${recM}:${recS < 10 ? '0' : ''}${recS}`;
  }

  _deletePlayer() {
    if (this.player) {
      this.player.destroy();
      this.player = null;
    }
  }

  _clearInactivePlayer() {
    if (this._progressInterval) {
      clearInterval(this._progressInterval);
      this._progressInterval = null;
    }
    if (this._playInterval) {
      clearInterval(this._playInterval);
      this._playInterval = null;
    }
    this.setState({ playButtonDisabled: false, playTime: 0, progress: 0, playerExists: false });
    this._deletePlayer();
  }

  _shouldUpdateProgressBar() {
    // Debounce progress bar update by 200 ms
    return Date.now() - this.lastSeek > 200;
  }

  _pause() {
    if (this.player) {
      this.player.pause((err) => {
        if (err) {
          AppUtil.debug(`ERROR IN PAUSING AUDIO PLAYER ${JSON.stringify(err)}`);
        }
        clearInterval(this._playInterval);
        this.setState({ playButtonDisabled: false });
      });
    }
  }

  _seek(percentage) {
    if (!this.player) {
      return;
    }

    this.setState({ loadingAudio: false });
    clearInterval(this._playInterval);
    this.lastSeek = Date.now();
    const position = percentage * this.player.duration;
    this.player.seek(position, () => {
      this._getPlayTimer();
    });
  }

  _getPlayTimer() {
    this._playInterval = setInterval(() => {
      if (this.player && this.player.currentTime < this.player.duration) {
        this.setState({ playTime: this.player.currentTime });
      } else {
        clearInterval(this._playInterval);
      }
    }, 1000);
  }

  _getProgressTimer() {
    this._progressInterval = setInterval(() => {
      if (this.player && this._shouldUpdateProgressBar() && this.player.duration > 0) {
        this.setState({
          progress: Math.max(0, this.player.currentTime) / this.player.duration,
        });
      }
    }, 100);
  }

  _startPlayer() {
    if (!this.player) {
      this.setState({ loadingAudio: true });

      if (!this._progressInterval) {
        this._getProgressTimer();
      }

      const { audioFile } = this.props;
      this.player = new Player(audioFile, {
        autoDestroy: false,
      });
      this.player.play(() => {
        this.setState({
          playButtonDisabled: true,
          duration: this.player.duration,
          playing: true,
          loadingAudio: false,
          playerExists: true,
        });
        this._getPlayTimer();
      });
    } else if (this.player && (this.player.canPrepare || this.player.canPlay)) {
      this.setState({ loadingAudio: true });
      this.player.play(() => {
        this.setState({
          playButtonDisabled: true,
          duration: this.player.duration,
          playing: true,
          loadingAudio: false,
        });
        this._getPlayTimer();
      });
    } else if (this.player && this.player.isPaused) {
      this.player.play(() => {
        this.setState({ playButtonDisabled: true, loadingAudio: false });
        this._getPlayTimer();
      });
    }

    this.player.on('ended', () => {
      clearInterval(this._playInterval);
      this.setState({
        playButtonDisabled: false,
        playTime: 0,
        progress: 0,
        playing: false,
      });
      // destroy player & clear intervals after 35sec
      setTimeout(() => {
        if (this.player && this.player.isStopped) {
          this._clearInactivePlayer();
        }
      }, 35000);
    });

    this.player.on('pause', () => {
      // destroy player & clear intervals after 35sec
      setTimeout(() => {
        if (this.player && this.player.isPaused) {
          this._clearInactivePlayer();
        }
      }, 35000);
    });

    this.player.on('error', () => {
      this._clearInactivePlayer();
    });
  }

  _deleteMessage() {
    const { deleteMessage } = this.props;
    Alert.alert(
      'Delete',
      'Do you want to delete message?',
      [
        { text: 'No', onPress: () => {}, style: 'cancel' },
        { text: 'Yes', onPress: () => deleteMessage() },
      ],
      { cancelable: false },
    );
  }

  render() {
    const {
      playButtonDisabled,
      progress,
      playTime,
      duration,
      playing,
      loadingAudio,
      playerExists,
    } = this.state;
    const { showPlayer, showDelete, position } = this.props;
    const playTimer = this.getTimerString(playTime);
    const durationTimer = this.getTimerString(duration);
    const disableSeek = playButtonDisabled || !playerExists;
    let audioPlayer;
    if (showPlayer) {
      audioPlayer = (
        <View style={styles.audioPlayerContainer}>
          <View style={styles.audioPlayerRow}>
            <View style={[styles.marginLeft15, styles.rowAlignJustifyCenter]}>
              {!playButtonDisabled &&
                !loadingAudio && (
                  <TouchableOpacity onPress={() => this._startPlayer()}>
                    <View
                      style={[
                        styles.audioPlayerButton,
                        {
                          backgroundColor:
                            position === 'right'
                              ? Colors.BG_CHAT_DETAIL_BUBBLE_RIGHT
                              : Colors.BG_CHAT_DETAIL_BUBBLE_LEFT,
                        },
                      ]}
                    >
                      <Icon
                        name="play-circle-outline"
                        type="material-community"
                        size={40}
                        color={
                          position === 'right' ? Colors.AUDIO_ICON_RIGHT : Colors.AUDIO_ICON_LEFT
                        }
                      />
                    </View>
                  </TouchableOpacity>
                )}
              {loadingAudio && (
                <View
                  style={[
                    styles.audioPlayerButton,
                    {
                      backgroundColor:
                        position === 'right'
                          ? Colors.BG_CHAT_DETAIL_BUBBLE_RIGHT
                          : Colors.BG_CHAT_DETAIL_BUBBLE_LEFT,
                    },
                  ]}
                >
                  <Icon
                    name="timer-sand"
                    type="material-community"
                    size={24}
                    color={position === 'right' ? Colors.AUDIO_ICON_RIGHT : Colors.AUDIO_ICON_LEFT}
                  />
                </View>
              )}
              {playButtonDisabled &&
                !loadingAudio && (
                  <TouchableOpacity onPress={() => this._pause()}>
                    <View
                      style={[
                        styles.audioPlayerButton,
                        {
                          backgroundColor:
                            position === 'right'
                              ? Colors.BG_CHAT_DETAIL_BUBBLE_RIGHT
                              : Colors.BG_CHAT_DETAIL_BUBBLE_LEFT,
                        },
                      ]}
                    >
                      <Icon
                        name="stop-circle-outline"
                        type="material-community"
                        size={40}
                        color={
                          position === 'right' ? Colors.AUDIO_ICON_RIGHT : Colors.AUDIO_ICON_LEFT
                        }
                      />
                    </View>
                  </TouchableOpacity>
                )}
            </View>
            <View style={styles.audioPlayerSliderContainer}>
              <Slider
                step={0.0001}
                disabled={disableSeek}
                onValueChange={(percentage) => this._seek(percentage)}
                value={progress}
                minimumTrackTintColor={Colors.AUDIO_ICON_RIGHT}
                thumbTintColor={
                  position === 'right' ? Colors.AUDIO_ICON_RIGHT : Colors.AUDIO_ICON_LEFT
                }
              />
            </View>
            {showDelete && (
              <TouchableOpacity
                style={styles.audioPlayerDeleteButton}
                onPress={() => this._deleteMessage()}
              >
                <FontAwesomeIcon
                  name="trash-o"
                  type="material-community"
                  color={position === 'right' ? Colors.AUDIO_ICON_RIGHT : Colors.AUDIO_ICON_LEFT}
                  size={28}
                />
              </TouchableOpacity>
            )}
          </View>

          <View style={styles.audioPlayerTimerContainer}>
            <Text
              numberOfLines={1}
              style={[
                styles.audioPlayerTimerText,
                {
                  color: position === 'right' ? Colors.TEXT_WHITE : Colors.TEXT_LEFT_TIME,
                },
              ]}
            >
              {playing ? playTimer : durationTimer}
            </Text>
          </View>
        </View>
      );
    } else {
      audioPlayer = <View style={[styles.audioPlayerContainer, { minHeight: 65 }]} />;
    }
    return audioPlayer;
  }
}

AudioPlay.defaultProps = {
  audioFile: '',
  showPlayer: false,
  showDelete: false,
  position: '',
  deleteMessage: () => {},
};

AudioPlay.propTypes = {
  audioFile: PropTypes.string,
  showPlayer: PropTypes.bool,
  showDelete: PropTypes.bool,
  position: PropTypes.string,
  deleteMessage: PropTypes.func,
};
