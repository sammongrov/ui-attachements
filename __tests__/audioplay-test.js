import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import AppUtil from '@utils';
import { Alert } from 'react-native';
import { Colors } from '@ui/theme_default';
import AudioPlay from '../AudioPlay';

configure({ adapter: new Adapter() });

jest.spyOn(Date, 'now').mockImplementation(() => 1479427300700);

jest.mock('@utils', () => ({
  debug: jest.fn(),
}));

jest.mock('Alert', () => ({
  alert: jest.fn((str1, str2, arr) => {
    arr[0].onPress();
  }),
}));

jest.mock('react-native-audio-toolkit', () => {
  function MockPlayer() {
    this.destroy = jest.fn();
    this.pause = jest.fn((cb) => {
      cb(null);
      return this;
    });
    this.play = jest.fn((cb) => {
      cb();
      return this;
    });
    this.seek = jest.fn((position, cb) => {
      cb();
      return this;
    });
    this.on = jest.fn((event, cb) => {
      if (event === 'ended') {
        cb();
      }
      return this;
    });
    this.duration = 369;
    this.currentTime = 12;
    this.canPrepare = true;
    this.canPlay = true;
    this.isPaused = false;
    this.isStopped = true;
  }
  return { Player: MockPlayer };
});

const audioFile = '//recordings/myVoice1.mp4';
const showPlayer = true;
const showDelete = true;
const position = 'left';
const deleteMessage = jest.fn();
const props = { audioFile, showPlayer, showDelete, position, deleteMessage };
const error = new Error('player error');

jest.useFakeTimers();

beforeEach(() => {
  jest.resetModules();
});

/* ------------------------- Snapshots ----------------------- */
it('AttachAudio renders correctly without props', () => {
  const tree = renderer.create(<AudioPlay />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('AttachAudio renders correctly with props', () => {
  const tree = renderer.create(<AudioPlay {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});

/* ------------------- component methods --------------------- */

it('AudioPlay willMount', () => {
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  expect(instance.player).toBeNull();
  expect(instance.lastSeek).toBe(0);
  expect(setInterval).toBeCalled();
});

it('AudioPlay willUnmount and clears a progress interval', () => {
  clearInterval.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance._progressInterval = 1;
  instance._playInterval = null;
  instance._deletePlayer = jest.fn();
  tree.unmount();
  expect(clearInterval).toBeCalledTimes(1);
  expect(instance._deletePlayer).toBeCalled();
});

it('AudioPlay willUnmount and clears a play interval', () => {
  clearInterval.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance._progressInterval = null;
  instance._playInterval = 2;
  instance._deletePlayer = jest.fn();
  tree.unmount();
  expect(clearInterval).toBeCalledTimes(1);
  expect(instance._deletePlayer).toBeCalled();
});

it('AudioPlayer calls getTimerString', () => {
  const timeInMS0 = 0;
  const timeInMS1 = 5000;
  const timeInMS2 = 42000;
  const timeInMS3 = 77000;
  const timeInMS4 = 723000;
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  const result0 = instance.getTimerString(timeInMS0);
  const result1 = instance.getTimerString(timeInMS1);
  const result2 = instance.getTimerString(timeInMS2);
  const result3 = instance.getTimerString(timeInMS3);
  const result4 = instance.getTimerString(timeInMS4);
  expect(result0).toMatch('00:00');
  expect(result1).toMatch('00:05');
  expect(result2).toMatch('00:42');
  expect(result3).toMatch('01:17');
  expect(result4).toMatch('12:03');
});

it('AudioPlayer calls _deletePlayer - no player', () => {
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = null;
  expect(instance._deletePlayer()).toBeUndefined();
});

it('AudioPlayer calls _deletePlayer - the player is present', () => {
  const destroy = jest.fn();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = { destroy };
  instance._deletePlayer();
  expect(destroy).toBeCalled();
  expect(instance.player).toBeNull();
});

it('AudioPlayer calls _clearInactivePlayer with a progress interval', () => {
  clearInterval.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance._progressInterval = 1;
  instance._playInterval = null;
  instance._deletePlayer = jest.fn();
  instance._clearInactivePlayer();
  expect(clearInterval).toBeCalledTimes(1);
  expect(tree.state().playButtonDisabled).toBe(false);
  expect(tree.state().playTime).toBe(0);
  expect(tree.state().progress).toBe(0);
  expect(tree.state().playerExists).toBe(false);
  expect(instance._deletePlayer).toBeCalled();
});

it('AudioPlayer calls _clearInactivePlayer with a play interval', () => {
  clearInterval.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance._progressInterval = null;
  instance._playInterval = 2;
  instance._deletePlayer = jest.fn();
  instance._clearInactivePlayer();
  expect(clearInterval).toBeCalledTimes(1);
  expect(instance._deletePlayer).toBeCalled();
});

it('AudioPlayer _shouldUpdateProgressBar returns true', () => {
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.lastSeek = 1479427200000;
  expect(instance._shouldUpdateProgressBar()).toBe(true);
});

it('AudioPlayer _shouldUpdateProgressBar returns false', () => {
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.lastSeek = 1479427300625;
  expect(instance._shouldUpdateProgressBar()).toBe(false);
});

it('AudioPlayer calls _pause - no player', () => {
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = null;
  expect(instance._pause()).toBeUndefined();
});

it('AudioPlayer calls _pause - error returned', () => {
  AppUtil.debug.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = {
    pause: jest.fn((cb) => {
      cb(error);
    }),
  };
  instance._pause();
  expect(AppUtil.debug).toBeCalled();
});

it('AudioPlayer calls _pause - the player paused', () => {
  clearInterval.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = {
    pause: jest.fn((cb) => {
      cb(null);
    }),
  };
  instance._pause();
  expect(clearInterval).toBeCalled();
  expect(tree.state().playButtonDisabled).toBe(false);
});

it('AudioPlayer calls _seek - no player', () => {
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = null;
  expect(instance._seek(0.35)).toBeUndefined();
});

it('AudioPlayer calls _seek with a player', () => {
  clearInterval.mockClear();
  const percentage = 0.35;
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = {
    seek: jest.fn((pos, cb) => {
      cb();
    }),
    duration: 15000,
  };
  instance._getPlayTimer = jest.fn();
  instance._seek(percentage);
  const playerPosition = percentage * instance.player.duration;
  expect(tree.state().loadingAudio).toBe(false);
  expect(clearInterval).toBeCalled();
  expect(instance.player.seek).toBeCalledWith(playerPosition, expect.any(Function));
});

it('AudioPlayer calls _getPlayTimer - no player', () => {
  clearInterval.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = null;
  instance._getPlayTimer();
  jest.advanceTimersByTime(1000);
  expect(clearInterval).toBeCalled();
});

it('AudioPlayer calls _getPlayTimer - a track is completed', () => {
  clearInterval.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = {
    currentTime: 15653,
    duration: 15653,
  };
  instance._getPlayTimer();
  jest.advanceTimersByTime(1000);
  expect(clearInterval).toBeCalled();
});

it('AudioPlayer calls _getPlayTimer - a track is ongoing', () => {
  clearInterval.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = {
    currentTime: 10810,
    duration: 15653,
  };
  instance._getPlayTimer();
  jest.advanceTimersByTime(1000);
  expect(tree.state().playTime).toBe(instance.player.currentTime);
  expect(clearInterval).not.toBeCalled();
});

it('AudioPlayer calls _getProgressTimer - no player', () => {
  setInterval.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = null;
  instance._getProgressTimer();
  jest.advanceTimersByTime(100);
  expect(setInterval).toBeCalled();
});

it('AudioPlayer calls _getProgressTimer - a progress bar should not be updated', () => {
  setInterval.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = {
    duration: 15653,
  };
  instance._shouldUpdateProgressBar = jest.fn(() => false);
  instance._getProgressTimer();
  jest.advanceTimersByTime(100);
  expect(setInterval).toBeCalled();
});

it('AudioPlayer calls _getProgressTimer - a duration == 0', () => {
  setInterval.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = {
    duration: 0,
  };
  instance._shouldUpdateProgressBar = jest.fn(() => true);
  instance._getProgressTimer();
  jest.advanceTimersByTime(100);
  expect(setInterval).toBeCalled();
});

it('AudioPlayer calls _getProgressTimer successfully', () => {
  setInterval.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = {
    duration: 15653,
    currentTime: 1250,
  };
  instance._shouldUpdateProgressBar = jest.fn(() => true);
  instance._getProgressTimer();
  const progress = Math.max(0, instance.player.currentTime) / instance.player.duration;
  jest.advanceTimersByTime(100);
  expect(setInterval).toBeCalled();
  expect(tree.state().progress).toBe(progress);
});

it('AudioPlayer calls _startPlayer - no player', () => {
  setTimeout.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = null;
  instance._progressInterval = null;
  instance._getProgressTimer = jest.fn();
  instance._getPlayTimer = jest.fn();
  instance._clearInactivePlayer = jest.fn();
  instance._startPlayer();
  expect(instance._getProgressTimer).toBeCalled();
  expect(tree.state().playButtonDisabled).toBe(false);
  expect(tree.state().duration).toBe(369);
  expect(tree.state().playing).toBe(false);
  expect(tree.state().loadingAudio).toBe(false);
  expect(tree.state().playerExists).toBe(true);
  expect(tree.state().playTime).toBe(0);
  expect(tree.state().progress).toBe(0);
  jest.runOnlyPendingTimers();
  expect(setTimeout).toBeCalled();
  expect(instance._clearInactivePlayer).toBeCalled();
});

it('AudioPlayer calls _startPlayer - progress interval set', () => {
  setTimeout.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = null;
  instance._progressInterval = 1;
  instance._getProgressTimer = jest.fn();
  instance._getPlayTimer = jest.fn();
  instance._startPlayer();
  expect(instance._getProgressTimer).not.toBeCalled();
});

it('AudioPlayer calls _startPlayer - player canPrepare', () => {
  setTimeout.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = {
    play: jest.fn((cb) => {
      cb();
    }),
    canPrepare: true,
    canPlay: true,
    duration: 10001,
    on: jest.fn((event, cb) => {
      if (event === 'ended') {
        cb();
      }
      return this;
    }),
    isStopped: false,
  };
  instance._getProgressTimer = jest.fn();
  instance._getPlayTimer = jest.fn();
  instance._clearInactivePlayer = jest.fn();
  instance._startPlayer();
  expect(tree.state().playButtonDisabled).toBe(false);
  expect(tree.state().duration).toBe(10001);
  expect(tree.state().playing).toBe(false);
  expect(tree.state().loadingAudio).toBe(false);
  expect(instance._getPlayTimer).toBeCalled();
  expect(tree.state().playTime).toBe(0);
  expect(tree.state().progress).toBe(0);
  jest.runOnlyPendingTimers();
  expect(setTimeout).toBeCalled();
  expect(instance._clearInactivePlayer).not.toBeCalled();
});

it('AudioPlayer calls _startPlayer - player canPlay', () => {
  setTimeout.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = {
    play: jest.fn((cb) => {
      cb();
    }),
    canPrepare: false,
    canPlay: true,
    duration: 10001,
    on: jest.fn((event, cb) => {
      if (event === 'pause') {
        cb();
      }
      return this;
    }),
    isPaused: false,
  };
  instance._getProgressTimer = jest.fn();
  instance._getPlayTimer = jest.fn();
  instance._clearInactivePlayer = jest.fn();
  instance._startPlayer();
  expect(tree.state().playButtonDisabled).toBe(true);
  expect(tree.state().duration).toBe(10001);
  expect(tree.state().playing).toBe(true);
  expect(tree.state().loadingAudio).toBe(false);
  expect(instance._getPlayTimer).toBeCalled();
  jest.runOnlyPendingTimers();
  expect(setTimeout).toBeCalled();
  expect(instance._clearInactivePlayer).not.toBeCalled();
});

it('AudioPlayer calls _startPlayer - player isPaused', () => {
  setTimeout.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = {
    play: jest.fn((cb) => {
      cb();
    }),
    canPrepare: false,
    canPlay: false,
    duration: 10001,
    on: jest.fn((event, cb) => {
      if (event === 'pause') {
        cb();
      }
      return this;
    }),
    isPaused: true,
  };
  instance._getProgressTimer = jest.fn();
  instance._getPlayTimer = jest.fn();
  instance._clearInactivePlayer = jest.fn();
  instance._startPlayer();
  expect(tree.state().playButtonDisabled).toBe(true);
  expect(tree.state().loadingAudio).toBe(false);
  expect(instance._getPlayTimer).toBeCalled();
  jest.runOnlyPendingTimers();
  expect(setTimeout).toBeCalled();
  expect(instance._clearInactivePlayer).toBeCalled();
});

it('AudioPlayer calls _startPlayer - player on error', () => {
  setTimeout.mockClear();
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance.player = {
    play: jest.fn((cb) => {
      cb();
    }),
    canPrepare: false,
    canPlay: false,
    duration: 10001,
    on: jest.fn((event, cb) => {
      if (event === 'error') {
        cb();
      }
      return this;
    }),
    isPaused: false,
  };
  instance._clearInactivePlayer = jest.fn();
  instance._startPlayer();
  expect(instance._clearInactivePlayer).toBeCalled();
});

it('VideoPlayer calls _deleteMessage, alert with "no" option', () => {
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance._deleteMessage();
  expect(Alert.alert).toBeCalled();
  expect(deleteMessage).not.toBeCalled();
});

it('VideoPlayer calls _deleteMessage, alert with "yes" option', () => {
  Alert.alert = jest.fn((str1, str2, arr) => {
    arr[1].onPress();
  });
  const tree = shallow(<AudioPlay {...props} />);
  const instance = tree.instance();
  instance._deleteMessage();
  expect(Alert.alert).toBeCalled();
  expect(deleteMessage).toBeCalled();
});

it('VideoPlayer calls _deleteMessage, alert with "yes" option, default prop', () => {
  Alert.alert = jest.fn((str1, str2, arr) => {
    arr[1].onPress();
  });
  const tree = shallow(
    <AudioPlay audioFile={audioFile} showPlayer={true} showDelete={true} position="right" />,
  );
  const instance = tree.instance();
  instance._deleteMessage();
  expect(Alert.alert).toBeCalled();
});

/* --------------------Simulations --------------------------- */
it('onPress start button', () => {
  const tree = shallow(<AudioPlay {...props} position="right" />);
  tree.setState({ playButtonDisabled: false, loadingAudio: false });
  const instance = tree.instance();
  instance._startPlayer = jest.fn();
  const playButton = tree.find('TouchableOpacity').first();
  playButton.props().onPress();
  expect(instance._startPlayer).toBeCalled();
});

it('loadingAudio', () => {
  const tree = shallow(<AudioPlay {...props} position="right" />);
  tree.setState({ loadingAudio: true });
  const icon = tree.find({ name: 'timer-sand' }).first();
  expect(icon.props().color).toBe(Colors.AUDIO_ICON_RIGHT);
});

it('onPress pause button', () => {
  const tree = shallow(<AudioPlay {...props} position="right" />);
  tree.setState({ playButtonDisabled: true, loadingAudio: false });
  const instance = tree.instance();
  instance._pause = jest.fn();
  const pauseButton = tree.find('TouchableOpacity').first();
  pauseButton.props().onPress();
  expect(instance._pause).toBeCalled();
});

it('onValueChange of a slider', () => {
  const tree = shallow(<AudioPlay {...props} position="right" />);
  const instance = tree.instance();
  instance._seek = jest.fn();
  const slider = tree.find('Slider').first();
  slider.props().onValueChange(0.78);
  expect(instance._seek).toBeCalledWith(0.78);
});

it('onValueChange of a slider', () => {
  const tree = shallow(<AudioPlay {...props} position="right" showDelete={true} />);
  const instance = tree.instance();
  instance._deleteMessage = jest.fn();
  const deleteButton = tree.find('TouchableOpacity').last();
  deleteButton.props().onPress();
  expect(instance._deleteMessage).toBeCalled();
});
