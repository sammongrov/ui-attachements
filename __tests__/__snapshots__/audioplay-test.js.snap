// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttachAudio renders correctly with props 1`] = `
<View>
  <View>
    <View
      style={
        Array [
          undefined,
          undefined,
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              undefined,
              Object {
                "backgroundColor": "#EFEFEF",
              },
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#888888",
                  "fontSize": 40,
                },
                undefined,
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            type="material-community"
          >
            
          </Text>
        </View>
      </View>
    </View>
    <View>
      <RCTSlider
        disabled={true}
        enabled={false}
        maximumValue={1}
        minimumTrackTintColor="#fff"
        minimumValue={0}
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onValueChange={[Function]}
        step={0.0001}
        style={
          Array [
            Object {
              "height": 40,
            },
            undefined,
          ]
        }
        thumbTintColor="#888888"
        value={0}
      />
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#888888",
              "fontSize": 28,
            },
            undefined,
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
        type="material-community"
      >
        
      </Text>
    </View>
  </View>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      numberOfLines={1}
      style={
        Array [
          undefined,
          Object {
            "color": "#888888",
          },
        ]
      }
    >
      00:00
    </Text>
  </View>
</View>
`;

exports[`AttachAudio renders correctly without props 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "minHeight": 65,
      },
    ]
  }
/>
`;
